pkg/health/membergarbagecollector/membergarbagecollector.go

```diff
@@ -0,0 +1,173 @@
+	}, nil
+}
+
+// RunMemberGarbageCollectorPeriodically periodically calls the member garbage collector
+func RunMemberGarbageCollectorPeriodically(ctx context.Context, hconfig *brtypes.HealthConfig, logger *logrus.Entry, etcdConfig *etcdutil.EtcdConnectionConfig) {
+
+	clientSet, err := miscellaneous.GetKubernetesClientSetOrError()
+	if err != nil {
+		logger.Errorf("failed to create kubernetes clientset: %v", err)
+		return
+	}
+	mgc, err := NewMemberGarbageCollector(logger, etcdConfig, clientSet)
+	if err != nil {
+		logger.Errorf("failed to initialize new heartbeat: %v", err)
+		return
+	}
+	mgc.gcTimer = time.NewTimer(hconfig.MemberGCDuration.Duration)
+	defer mgc.gcTimer.Stop()
+	mgc.logger.Info("Started etcd member garbage collection timer")
+
+	for {
+		select {
+		case <-mgc.gcTimer.C:
+			err := mgc.RemoveSuperfluousMembers(ctx)
+			if err != nil {
+				mgc.logger.Warn(err)
+			}
+			mgc.gcTimer.Reset(hconfig.MemberGCDuration.Duration)
+		case <-ctx.Done():
+			mgc.logger.Info("Stopped etcd member garbage collection timer")
+			return
+		}
+	}
+}
+
+// RemoveSuperfluousMembers removes a member from the etcd cluster if its corresponding pod is not present
+func (mgc *MemberGarbageCollector) RemoveSuperfluousMembers(ctx context.Context) error {
+	etcd := &druidv1alpha1.Etcd{}
+	if err := mgc.k8sClient.Get(ctx, client.ObjectKey{
+		Namespace: mgc.podNamespace,
+		Name:      mgc.etcdName,
+	}, etcd); err != nil {
+		return &errors.EtcdError{
+			Message: fmt.Sprintf("Could not fetch etcd %v with error: %v", mgc.etcdName, err),
+		}
+	}
+	// Return is no members present
+	if len(etcd.Status.Members) < 1 {
+		return &errors.EtcdError{
+			Message: "There are no members present in the etcd status. Nothing to clean",
+		}
+	}
+
+	// Return if memberlist size <= desired cluster size. Cleanup not needed
+	if len(etcd.Status.Members) <= etcd.Spec.Replicas {
+		return nil
+	}
+
+	//Create etcd client to get etcd cluster member list
+	etcdClient, err := etcdutil.GetTLSClientForEtcd(mgc.etcdConfig)
+	if err != nil {
+		return &errors.EtcdError{
+			Message: fmt.Sprintf("Failed to create etcd client: %v", err),
+		}
+	}
+	defer etcdClient.Close()
+
+	etcdMemberListResponse, err := etcdClient.MemberList(ctx)
+	if err != nil {
+		return &errors.EtcdError{
+			Message: fmt.Sprintf("Could not list etcd cluster members: %v", err),
+		}
+	}
+	for _, x := range etcdMemberListResponse.Members {
+		memberPod := &v1.Pod{}
+		err := mgc.k8sClient.Get(ctx, client.ObjectKey{
+			Namespace: mgc.podNamespace,
+			Name:      x.Name,
+		}, memberPod)
+		if apierrors.IsNotFound(err) {
+			//Remove member from etcd cluster if corresponding pod cannot be found
+			mgc.logger.Info("Removing superfluous member ", x.ID, " : ", err)
+			if _, err = etcdClient.MemberRemove(ctx, x.ID); err != nil {
+				return &errors.EtcdError{
+					Message: fmt.Sprintf("Error removing member %v from the cluster: %v", x.ID, err),
+				}
+			}
+		}
```

Comment:
Here the member is removed if the corresponding pod doesn't exist. But `etch-druid` removes the member from the status based more checks (like `lease` etc.). So, it might be better to be fully sure by checking that the lease, `Etcd` `status.members` and `pods` all agree that this member is now gone. To be precise, if you check the `status.members` and the pods then checking the `leases` may not be required because the `status.members` is kept in sync with the `leases`.

Reply:
Done